chain/network/src/snapshot_hosts/mod.rs

```diff
@@ -133,4 +312,37 @@ impl SnapshotHostsCache {
     pub fn get_hosts(&self) -> Vec<Arc<SnapshotHostInfo>> {
         self.0.lock().hosts.iter().map(|(_, v)| v.clone()).collect()
     }
+
+    // Selecs a peer to send the request for this part ID to. Chooses based on a priority score
+    // calculated as a hash of the Peer ID plus the part ID, and will return different hosts
+    // on subsequent calls, eventually iterating over all valid SnapshotHostInfos we know about
+    // TODO: get rid of the dead_code and hook this up to the decentralized state sync
+    #[allow(dead_code)]
+    pub fn select_host(
+        &self,
+        sync_hash: &CryptoHash,
+        shard_id: ShardId,
+        part_id: &PartId,
+    ) -> Option<PeerId> {
+        let mut inner = self.0.lock();
+        let num_hosts = inner.hosts.len();
+        let selector = inner.state_part_selectors.entry(shard_id).or_default();
+
+        if selector.tried_everybody(part_id) && selector.len(part_id) < num_hosts {
```

Comment:
Just to check my understanding: is this essentially a memory optimization to avoid duplicating the entire set of hosts into a ` PartPeerSelector` for each part?

Reply:
yup, it seems like it would be kinda wasteful to store them all up front if the first few will do. Maybe also in the future we could drop ones that are clearly not responding, maybe also storing that on disk so we don't try again after a restart